File: src/DynamicLinkLibraries/Motion6D/Motion6D.Portable/RigidReferenceFrame.cs
using System;
using System.Collections.Generic;
using System.Text;


using CategoryTheory;
using DataPerformer.Interfaces;
using Diagram.UI.Interfaces;
using Motion6D.Interfaces;
using NamedTree;
using RealMatrixProcessor;
using Vector3D;

namespace Motion6D.Portable
{
    /// <summary>
    /// Rigid reference frame
    /// </summary>
    public class RigidReferenceFrame : CategoryObject, IReferenceFrame,
       IPostLoadPosition, IPostSetArrow, IAlias, IStarted
    {

        #region Fields

        protected Vector3DProcessor vp = new();

        protected RealMatrix realMatrix = new();

        protected ReferenceFrame own = new Motion6DAcceleratedFrame();

        private ReferenceFrame relative = new Motion6DAcceleratedFrame();

        /// <summary>
        /// Relative position
        /// </summary>
        protected double[] relativePosition = new double[] { 0, 0, 0 };

        /// <summary>
        /// Relarive quaternion components
        /// </summary>
        protected double[] relativeQuaternion = new double[] { 1, 0, 0, 0 };

        List<IPosition> children = new List<IPosition>();

        /// <summary>
        /// Parent frame
        /// </summary>
        protected IReferenceFrame parent;
   
        /// <summary>
        /// Associated parameters
        /// </summary>
        protected object parameters;

        //protected double[,] relativeMatrix = new double[3, 3];

        /// <summary>
        /// Auxiliary variable
        /// </summary>
        protected double[,] q44 = new double[4, 4];


        /// <summary>
        /// Linear velocity
        /// </summary>
        protected double[] velocity = new double[] { 0, 0, 0 };

        // protected double[] relativeVelocity = new double[] { 0, 0, 0 };

        /// <summary>
        /// Angular velocity
        /// </summary>
        protected double[] omega = new double[] { 0, 0, 0 };

        protected string[] aliasNames = new string[] { "X", "Y", "Z", "Roll", "Pitch", "Yaw" };

        protected Dictionary<string, int> alinames = new Dictionary<string, int>();

        #endregion

        #region Ctor

        /// <summary>
        /// Default constructor
        /// </summary>
        public RigidReferenceFrame()
        {
            for (int i = 0; i < aliasNames.Length; i++)
            {
                alinames[aliasNames[i]] = i;
            }
            Init();
        }


        protected RigidReferenceFrame(object obj)
        {

        }


        #endregion

        #region IReferenceFrame Members

        /// <summary>
        /// Own frame
        /// </summary>
        public ReferenceFrame Own
        {
            get { return own; }
        }

  

        #endregion

        #region IStarted Members

        void IStarted.Start(double time)
        {
            Start(time);
        }

        #endregion

        #region IAlias Members

        IList<string> IAlias.AliasNames => aliasNames;

        object IAlias.this[string name] { get => GetAliasValue(name); set => SetAliasValue(name, value); }

        event Action<IAlias, string> change;

        event Action<IAlias, string> IAlias.OnChange
        {
            add
            {
                change += value;
            }

            remove
            {
                change -= value;
            }
        }

        event Action<IPosition> INode<IPosition>.OnAdd
        {
            add
            {
            }

            remove
            {
            }
        }

        event Action<IPosition> INode<IPosition>.OnRemove
        {
            add
            {
            }

            remove
            {
            }
        }

        object IAlias.GetType(string name)
        {
            return GetAliasType(name);
        }

        #endregion

        #region IAlias Support


        EulerAngles angles = new EulerAngles();

        protected virtual object GetAliasValue(string name)
        {
            int i = alinames[name];
            if (i < 3)
            {
                return relative.Position[i];
            }
            vp.Set(angles, relative.Quaternion);
            switch (i - 3)
            {
                case 0: return angles.roll;
                case 1: return angles.pitch;
                case 2: return angles.yaw;
            }
            return 0;
        }


        protected virtual void SetAliasValue(string name, object value)
        {
            double v = (double)value;
            int i = alinames[name];
            if (i < 3)
            {
                relative.Position[i] = v;
            }
            if (i < 9)
            {
                vp.Set(angles, relative.Quaternion);
                switch (i - 6)
                {
                    case 0:
                        angles.roll = v;
                        break;
                    case 1:
                        angles.pitch = v;
                        break;
                    case 2:
                        angles.yaw = v;
                        break;
                }
                vp.ToQuaternion(angles, relative.Quaternion);
            }
        }


        protected virtual object GetAliasType(string name)
        {
            return (double)0;
        }


        #endregion

        #region IPosition Members

        double[] IPosition.Position
        {
            get { return own.Position; }
        }

        /// <summary>
        /// Parent frame
        /// </summary>
        public virtual IReferenceFrame Parent
        {
            get
            {
                return parent;
            }
            set
            {
                if (value != null & parent != null)
                {
                    throw new ErrorHandler.OwnException("Parent");
                }
                parent = value;
                if (value == null)
                {
                    owp = Motion6DFrame.Base;
                    return;
                }
                if (IsSerialized)
                {
                    return;
                }
            }
        }

        /// <summary>
        /// Position parameters
        /// </summary>
        public virtual object Parameters
        {
            get
            {
                return parameters;
            }
            set
            {
                parameters = value;
            }
        }


        /// <summary>
        /// Updates itself
        /// </summary>
        public virtual void Update()
        {
            ReferenceFrame b = BaseFrame;
            own.Set(b, relative);
        }

        #endregion

        #region IPostLoadPosition Members

        /// <summary>
        /// The operation that performs after arrows setting
        /// </summary>
        public virtual void PostLoadPosition()
        {
            CreateFrame();
            CopyPositionToRelativeFrame();
            CopyQuaternionToRelativeFrame();
            Init();
            relative.SetMatrix();
        }


        #endregion

        #region IPostSetArrow Members

        /// <summary>
        /// The operation that performs after arrows setting
        /// </summary>
        public virtual void PostSetArrow()
        {
            PostSetParameters();
            CreateFrame();
        }

        #endregion

        #region Specific Members

        #region Public Members


        /// <summary>
        /// Copies position to relative reference frame
        /// </summary>
        public void CopyPositionToRelativeFrame()
        {
            if (IsSerialized)
            {
                return;
            }
            Array.Copy(relativePosition, relative.Position, 3);
        }


        /// <summary>
        /// Copies quaternion to relative reference frame
        /// </summary>
        public virtual void CopyQuaternionToRelativeFrame()
        {
            if (IsSerialized)
            {
                return;
            }
            Array.Copy(relativeQuaternion, relative.Quaternion, 4);
            relative.SetMatrix();
        }

        /// <summary>
        /// Copies whole 6D position to relative reference frame
        /// </summary>
        public void Copy6DPosition()
        {
            CopyPositionToRelativeFrame();
            CopyQuaternionToRelativeFrame();
        }

        /// <summary>
        /// Relative position
        /// </summary>
        public double[] RelativePosition
        {
            get { return relativePosition; }
        }

        /// <summary>
        /// Relative matrix
        /// </summary>
        public virtual double[,] RelativeMatrix
        {
            get
            {
                return relative.Matrix;
            }
            set
            {
                vp.MatrixToQuaternion(value, relativeQuaternion);
                CopyQuaternionToRelativeFrame();
            }
        }

        /// <summary>
        /// Initialization
        /// </summary>
        public void Init()
        {
            if (relative == null)
            {
                return;
            }
            double[] q = relative.Quaternion;
            for (int i = 0; i < q.Length; i++)
            {
                q[i] = relativeQuaternion[i];
            }
            double[] p = relative.Position;
            for (int i = 0; i < p.Length; i++)
            {
                p[i] = relativePosition[i];
            }
        }

        #endregion

        #region Protected Members

        /// <summary>
        /// The "is serialized" sign
        /// </summary>
        protected bool IsSerialized
        {
            get;
            set;
        } = false;


        /// <summary>
        /// Post set parameteres operation
        /// </summary>
        protected virtual void PostSetParameters()
        {
            Parameters = parameters;
        }



        /// <summary>
        /// Base frame
        /// </summary>
        protected ReferenceFrame BaseFrame
        {
            get
            {
                if (parent != null)
                {
                    return parent.Own;
                }
                return Motion6DFrame.Base;
            }
        }

        /// <summary>
        /// Creates frame
        /// </summary>
        protected virtual void CreateFrame()
        {
            if (IsAcceleration)
            {
                relative = new Motion6DAcceleratedFrame();
                owp = new Motion6DAcceleratedFrame();
            }
            else if (IsVelocity & IsAngularVelocity)
            {
                relative = new Motion6DFrame();
                owp = new Motion6DFrame();
            }
            else if (IsAngularVelocity)
            {
                relative = new RotatedFrame();
                owp = new RotatedFrame();
            }
            else if (IsVelocity)
            {
                relative = new MovedFrame();
                owp = new MovedFrame();
            }
            else
            {
                relative = new ReferenceFrame();
                owp = new ReferenceFrame();
            }
        }

        /// <summary>
        /// Relative reference frame
        /// </summary>
        protected ReferenceFrame Relative
        {
            get
            {
                return relative;
            }
            set
            {
                relative = value;
            }
        }

        protected void ClearAliases()
        {
            aliasNames = new string[0];
            alinames.Clear();
        }
       
        /// <summary>
        /// Start
        /// </summary>
        /// <param name="time">Time</param>
        protected virtual void Start(double time)
        {
            Update();
        }

        void INode<IPosition>.Add(INode<IPosition> node)
        {
            var position = node as IPosition;
            children.Add(position);
        }

        void INode<IPosition>.Remove(INode<IPosition> node)
        {
            var position = node as IPosition;
            children.Remove(position);
        }

        /// <summary>
        /// Own frame
        /// </summary>
        protected ReferenceFrame owp
        {
            set
            {
                own = value;
            }
        }

        /// <summary>
        /// The "is acceleration" sign
        /// </summary>
        protected virtual bool IsAcceleration
        {
            get
            {
                if (parent == null)
                {
                    return true;
                }
                return parent.Own is IAcceleration;
            }
        }

        /// <summary>
        /// Detects velocity support of ReferenceFrame class
        /// </summary>
        protected virtual bool IsVelocity
        {
            get
            {
                if (parent == null) // If parent frame is null
                {
                    return true;
                }
                return parent.Own is IVelocity; // Parent frame implements IVelocity interface
            }
        }

        /// <summary>
        /// The "is angular velocity" sign
        /// </summary>
        protected virtual bool IsAngularVelocity
        {
            get
            {
                if (parent == null)
                {
                    return true;
                }
                return parent.Own is IAngularVelocity;
            }
        }

        #endregion

        #region Internal Members

        internal ReferenceFrame RelativeFrame => relative;

        INode<IPosition> INode<IPosition>.Parent { get => this.Parent; set => Parent = value as IReferenceFrame; }
   
        IPosition INode<IPosition>.Value => this;

        /// <summary>
        /// Children objects
        /// </summary>
        IEnumerable<INode<IPosition>> INode<IPosition>.Nodes { get => children; set { } }

    
        #endregion

        #endregion

    }
}

File: src/TypeScript/TypeScriptLibrary/src/Library/Motion6D/Objects/RigidReferenceFrame.ts
import { CategoryObject } from "../../CategoryObject";
import { OwnError } from "../../ErrorHandler/OwnError";
import { Motion6DAcceleratedFrame } from "../Motion6DAcceleratedFrame";
import { Motion6DPerformer } from "../Motion6DPerformer";
import { ReferenceFrame } from "../ReferenceFrame";
import { Motion6DFrame } from "../Motion6DFrame";
import { RotatedFrame } from "../RotatedFrame";
import { MovedFrame } from "../MovedFrame";
import { RealMatrix } from "../../RealMatrixProcessor/RealMatrix";
import { Vector3DProcessor } from "../../Vector3D/Vector3DProcessor";
import { PerformerMeasuremets } from "../../Measurements/PerformerMeasuremets";
import type { IDesktop } from "../../Interfaces/IDesktop";
import type { INodeT } from "../../NamedTree/Interfaces/INodeT";
import type { IPosition } from "../Interfaces/IPosition";
import type { IReferenceFrame } from "../Interfaces/IReferenceFrame";
import type { IPostLoadPosition } from "../Interfaces/IPostLoadPosition";
import type { IAlias } from "../../Interfaces/IAlias";
import type { IPostSetArrow } from "../../Interfaces/IPostSetArrow";


export class RigidReferenceFrame extends CategoryObject implements IReferenceFrame, IPostLoadPosition, IAlias, IPostSetArrow
{

    constructor(desktop: IDesktop, name: string) {
        super(desktop, name);
        this.typeName = "RigidReferenceFrame";
        this.types.push("IReferenceFrame");
        this.types.push("IPosition");
        this.types.push("IPostLoadPosition");
        this.types.push("IPostSetArrow");
        this.types.push("IAlias");
        this.types.push("RigidReferenceFrame");
    }

    



    /// </summary>
    protected relativePosition: number[] = [0, 0, 0];

    /// <summary> : IFunc<any>[] = [];
    /// Relarive quaternion components
    /// </summary>
    protected relativeQuaternion: number[] = [1, 0, 0, 0];

    children: IPosition[] = [];

    /// <summary>
    /// Parent frame
    /// </summary>
    protected parent!: IReferenceFrame;

    protected parentNode !: INodeT<IPosition>

    /// <summary>
    /// Associated parameters
    /// </summary>
    protected parameters: any;

    //protected double[,] relativeMatrix = new double[3, 3];
    /// <summary>
    /// Auxiliary variable
    /// </summary>
    protected q44: number[][] = [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]];


    /// <summary>
    /// Linear velocity
    /// </summary>
    protected velocity: number[] = [0, 0, 0, 0];

    // protected double[] relativeVelocity = new double[] { 0, 0, 0 };
    /// <summary>
    /// Angular velocity
    /// </summary>
    protected omega: number[] = [0, 0, 0];

    protected aliasNames: string[] = ["X", "Y", "Z", "Roll", "Pitch", "Yaw"];

    protected alinames: Map<string, number> = new Map<string, number>();


    protected vp: Vector3DProcessor = new Vector3DProcessor();

    protected realMatrix: RealMatrix = new RealMatrix();

    protected own: ReferenceFrame = new Motion6DAcceleratedFrame();

    protected relative: ReferenceFrame = new Motion6DAcceleratedFrame();

    protected mPerformer: Motion6DPerformer = new Motion6DPerformer();


    protected measuremrntPerformrer: PerformerMeasuremets = new PerformerMeasuremets();

    nodes: INodeT<IPosition>[] = [];

    postSetArrow(): void {
        this.setParameters(this.parameters)
        this.createFrame();
        this.relative.copyReferenceFrameFromPositionQuatetnion(this.relativePosition, this.relativeQuaternion)
        this.own.copyReferenceFrameFromPositionQuatetnion(this.relativePosition, this.relativeQuaternion)
    }


    getAliasNames(): string[] {
        return this.aliasNames;
    }
    getAliasType(name: string) {
        return 0;
    }
    getAliasValue(name: string) {

        return this.alinames.get(name);
    }
    setAliasValue(name: string, value: any): void {
        this.alinames.set(name, value)
    }

    postLoadPosition(): void {
        this.createFrame();
        this.copyPositionToRelativeFrame();
        this.copyQuaternionToRelativeFrame();
        this.init();
        this.relative.setMatrix();
      }

    protected impl(s: string): boolean {
        if (this.parent === undefined) { return true; }
        var own = this.parent.getOwnFrame();
        if (own === undefined) { return false }
        return this.performer.implementsType(own, s)
    }

    protected isAcceleration(): boolean {
        return this.impl("IAcceleration")
    }

    protected isVelocity(): boolean {
        return this.impl("IVelocity")
    }


    protected isAngularVelocity(): boolean {
        return this.impl("IAngularVelocity")
    }

    public copyPositionToRelativeFrame(): void {
        var rp = this.relative.getPosition();
        this.performer.copyArray<number>(this.relativePosition, rp);

    }

    public copyQuaternionToRelativeFrame(): void {
        var rp = this.relative.getQuaternion();
        this.performer.copyArray<number>(this.relativeQuaternion, rp);
        this.relative.setMatrix();
    }

    public copy6DPosition() : void {
        this.copyPositionToRelativeFrame();
        this.copyQuaternionToRelativeFrame();
    }

    public init(): void {
        if (this.relative === undefined) {
            return;
        }
        let q = this.relative.getQuaternion()
        for (let i = 0; i < q.length; i++)
        {
            q[i] = this.relativeQuaternion[i];
        }
        var p = this.relative.getPosition();
        for (let i = 0; i < p.length; i++)
        {
            p[i] = this.relativePosition[i];
        }
    }


    protected createFrame(): void {
        if (this.isAcceleration()) {
            this.relative = new Motion6DAcceleratedFrame();
            this.own = new Motion6DAcceleratedFrame();
        }
        if (this.isVelocity() && this.isAngularVelocity()) {
            this.relative = new Motion6DFrame();
            this.own = new Motion6DFrame();
        }
        if (this.isAngularVelocity()) {
            this.relative = new RotatedFrame();
            this.own = new RotatedFrame();
        }
        if (this.isVelocity()) {
            this.relative = new MovedFrame();
            this.own = new MovedFrame();
        }
        this.relative = new ReferenceFrame();
        this.own = new ReferenceFrame();
    }

    getNodeValueT(): IPosition {
        return this;
    }
    getParentT(): INodeT<IPosition> | undefined {
        return this.parentNode;
    }
    setParentT(parent: INodeT<IPosition>): void {
        this.parentNode = parent;
    }
    getNodesT(): INodeT<IPosition>[] {
        return this.nodes;
    }
    addNodeT(node: INodeT<IPosition>): void {
        this.nodes.push(node);
    }
    removeNodeT(node: INodeT<IPosition>): void {
        this.nodes = this.performer.remove(this.nodes, node)
    }
    getOwnFrame(): ReferenceFrame {
        return this.own;
    }
    getPosition(): number[] {
        return this.own.getPosition();
    }
    getParentFrame(): IReferenceFrame | undefined {
        return this.parent;
    }
    setParentFrame(parent: IReferenceFrame): void {
        if ((parent != undefined) && this.parent != undefined) {
            throw new OwnError("Parent", "", "");
        }
        this.parent = parent;
        if (parent == undefined) {
            this.own = this.mPerformer.getBaseFrame();
            return;
        }
    }
    getParameters(): any {
        return this.parameters;
    }

    setParameters(parameters: any): void {
        this.parameters = parameters;
    }

    updateReferenceFrame(): void {
        let own = this.getOwnFrame()
        let b = this.getBaseFrame();
        if (b === undefined)
        {
            this.relative.copyReferenceFrameFrom(own)
        }
        else
        {
            own.setReferenceFrame(b as ReferenceFrame, this.relative);
        }
    }

    protected getBaseFrame(): ReferenceFrame | undefined {
        if (this.parent === undefined) {
            return undefined;
        }
        else {
            return this.parent.getOwnFrame();
        }

    }
}